### ISSUE TITLE
Should there exist an empty KDTree?

### ISSUE THREAD

**Issue by JoshChristie** [KristofferC/NearestNeighbors.jl#35]
On julia 0.4 I found it useful to define an empty KDTree with 0 points (i.e.`tree = KDTree(rand(3,0))`). 
Is there currently a way to do this with the new changes? Is this the right thing to be doing?

**Comment by KristofferC** [comment]
If can just remove the assert and it works:

```
julia> KDTree(rand(3,0))
NearestNeighbors.KDTree{StaticArrays.SVector{3,Float64},Distances.Euclidean,Float64}
  Number of points: 0
  Dimensions: 3
  Metric: Distances.Euclidean()
  Reordered: true
```

 but could you explain to me the use of a 0 point tree. Perhaps I can fix it in a better way then.

**Comment by JoshChristie** [comment]
I have a [PointCloud](https://github.com/FugroRoames/PointClouds.jl) data structure that holds positions and a spatial index (KDTree). Removing (filtering) points from the PointCloud will recompute the KDTree. If the PointCloud is filtered down to 0 points it will have a 0 point tree.

**Comment by c42f** [comment]
Imagine you're processing point cloud data in tiles.  Most tiles contain large amounts of point cloud data, but some contain nothing (eg, weren't actually scanned).  However, you might not know exactly the detail of which tiles contain data until you go looking in a database, applying some filters, etc.

Now imagine that you've got a function which sets up data structures for a tile.  If the data structure contains a KDTree, you now can't instantiate it in the special case that zero points come back.  So you're reduced to throwing an error, returning an error code rather than a valid data structure or possibly leaving the field `#undef` (ugh!).  That's somewhat ok if you can do an early bailout on this tile, but if there's other things you might want to do with the tile (eg, other kinds of data present) you'll need a special case placeholder in the data structure - a "null tree" or some such.  This will break type stability a bit and introduce unnecessary special cases.

(Special case code is particularly bad when you're processing a lot of data, since every special case happens eventually.  If you've got N special cases which might occur, you end up with N! combinations of those.  So I think it's better to make the special case equal to the general case where possible.)

**Comment by KristofferC** [comment]
Yeah, I get it. I can just special case for 0 points.

**Comment by c42f** [comment]
Great, I was sure you'd get it, sorry if that was a bit over-explained.

So I take it that this is not as simple as removing the assert and some other things will break in special ways inside `KDTree`?

**Comment by KristofferC** [comment]
Do you also want to query things for the empty tree? Like `inrange` and get back empty lists etc?

**Comment by c42f** [comment]
Good question, I think this ideally should work as a general matter of avoiding special cases.

**Comment by c42f** [comment]
@JoshChristie - what do you need for this right now?

**Comment by JoshChristie** [comment]
Removing the assert is sufficient. 

The current behaviour of `inrange` will return an empty list, which makes sense. 

`knn` gives a `ERROR: ArgumentError: k > number of points in tree or ≦ 0`. Should it also get back an empty list?

**Comment by KristofferC** [comment]
Perhaps 0 should be allowed and give empty lists.

**Comment by c42f** [comment]
> Perhaps 0 should be allowed and give empty lists

Seems reasonable.

Should `knn` be allowed to return `< k` points for small trees, or is this just a bad idea?  On the one hand, I'd naively expect `knn` to return exactly `k` points.  On the other hand, when writing robust code with an unknown tree size I'd be forced to play tricks like:

```julia
points = randn(3,10)
tree = KDTree(points)
# ...
k = min(10, size(points,2))
neighbors,dists = knn(tree, mypoint, k)
```

Hmm, this is not that bad actually.  At least it's explicit that `k` might be "smaller than expected".

**Comment by KristofferC** [comment]
Closed by #36
